cpp/src/arrow/compute/row/encode_internal.h

```diff
@@ -0,0 +1,323 @@
+                             const std::vector<KeyColumnArray>& cols);
+  /// \brief Encode selection of prepared rows into a row table
+  /// \param rows The output row table
+  /// \param num_selected The number of rows to encode
+  /// \param selection indices of the rows to encode
+  Status EncodeSelected(RowTableImpl* rows, uint32_t num_selected,
+                        const uint16_t* selection);
+
+  /// \brief Decode a window of row oriented data into a corresponding
+  ///        window of column oriented storage.
+  /// \param start_row_input The starting row to decode
+  /// \param start_row_output An offset into the output array to write to
+  /// \param num_rows The number of rows to decode
+  /// \param rows The row table to decode from
+  /// \param cols The columns to decode into, should be sized appropriately
+  ///
+  /// The output buffers need to be correctly allocated and sized before
+  /// calling each method.  For that reason decoding is split into two functions.
+  /// DecodeFixedLengthBuffers processes everything except for varying length
+  /// buffers.
+  /// The output can be used to find out required varying length buffers sizes
+  /// for the call to DecodeVaryingLengthBuffers
+  void DecodeFixedLengthBuffers(int64_t start_row_input, int64_t start_row_output,
+                                int64_t num_rows, const RowTableImpl& rows,
+                                std::vector<KeyColumnArray>* cols);
+
+  /// \brief Decode the varlength columns of a row table into column storage
+  /// \param start_row_input The starting row to decode
+  /// \param start_row_output An offset into the output arrays
+  /// \param num_rows The number of rows to decode
+  /// \param rows The row table to decode from
+  /// \param cols The column arrays to decode into
+  void DecodeVaryingLengthBuffers(int64_t start_row_input, int64_t start_row_output,
+                                  int64_t num_rows, const RowTableImpl& rows,
+                                  std::vector<KeyColumnArray>* cols);
+
+ private:
+  /// Prepare column array vectors.
+  /// Output column arrays represent a range of input column arrays
+  /// specified by starting row and number of rows.
+  /// Three vectors are generated:
+  /// - all columns
+  /// - fixed-length columns only
+  /// - varying-length columns only
+  void PrepareKeyColumnArrays(int64_t start_row, int64_t num_rows,
+                              const std::vector<KeyColumnArray>& cols_in);
+
+  LightContext* ctx_;
+
+  // Data initialized once, based on data types of key columns
+  RowTableMetadata row_metadata_;
+
+  // Data initialized for each input batch.
+  // All elements are ordered according to the order of encoded fields in a row.
+  std::vector<KeyColumnArray> batch_all_cols_;
+  std::vector<KeyColumnArray> batch_varbinary_cols_;
+  std::vector<uint32_t> batch_varbinary_cols_base_offsets_;
+};
+
+class EncoderInteger {
```

Comment:
Do these all have to be exposed in a `.h`?

Reply:
Some don't.  Any of the encoders that have an `AVX2` implemented method do I think.  So if I was going to need an internal header anyways it seemed more consistent to just throw them all in.  However, I can prune this down to just the encoders needed if that would be better.